chunky/src/java/se/llbit/chunky/renderer/scene/SkyCache.java

```diff
@@ -16,88 +16,101 @@
-   * requests will pull from the cache.
-   *
-   * Default cache size is 1024x1024 with bilinear interpolation which seems to work well for all simulated sky
-   * modes.
+   * An on-the-fly sky cache. Automatically calculates sky colors as they are requested and caches
+   * them. Any repeat requests will pull from the cache.
+   * <p>
+   * Default cache size is 1024x1024 with bilinear interpolation which seems to work well for all
+   * simulated sky modes.
    *
    * @param sky Sky object to pull sky renderer from.
    */
   public SkyCache(Sky sky) {
-    this.sky = sky;
-    skyTexture = null;
+    simSky = sky.getSimulatedSky();
+    precalculateSky();
   }
 
-  /** Sync this cache with another cache */
-  public void syncCache(SkyCache cache) {
+  /**
+   * Set this cache's content to the content of another cache.
+   */
+  public synchronized void set(SkyCache cache) {
     this.skyResolution = cache.skyResolution;
     this.simSky = cache.simSky;
-    skyTexture = null;
+    this.skyTexture = cache.skyTexture;
   }
 
-  /** Reset the sky cache */
-  public void reset(Sky sky) {
-    simSky = sky.getSimulatedSky();
-
-    // Sky has not yet been initialized. No need to reset.
-    if (skyTexture == null) {
-      return;
-    }
-
-    for (int i = 0; i < skyResolution+1; i++) {
-      for (int j = 0; j < skyResolution+1; j++) {
-        for (int k = 0; k < 3; k++) {
-          skyTexture[i][j][k] = -1;
+  /**
+   * Fill the sky cache
+   */
+  public synchronized void precalculateSky() {
+    double[][][] skyTexture = new double[skyResolution + 1][skyResolution + 1][3];
+
+    ForkJoinPool pool = new ForkJoinPool(PersistentSettings.getNumThreads());
+    pool.submit(() -> {
+      IntStream.range(0, skyResolution + 1).parallel().forEach(i -> {
+        for (int j = 0; j < skyResolution + 1; j++) {
+          Vector3 c = getSkyColorAt(i, j);
+          skyTexture[i][j][0] = c.x;
+          skyTexture[i][j][1] = c.y;
+          skyTexture[i][j][2] = c.z;
         }
-      }
-    }
+      });
+    }).join();
+    pool.shutdownNow();
+
+    this.skyTexture = skyTexture;
   }
 
-  /** Adjust the sky resolution and reset the cache */
+  /**
+   * Adjust the sky resolution and reset the cache
+   */
   public void setSkyResolution(int skyResolution) {
     this.skyResolution = skyResolution;
-    skyTexture = null;
+    precalculateSky();
   }
 
-  /** Get the current sky resolution */
+  /**
+   * Get the current sky resolution
+   */
   public int getSkyResolution() {
     return this.skyResolution;
   }
 
+  public void setSimulatedSkyMode(SimulatedSky skyMode) {
+    this.simSky = skyMode;
+    precalculateSky();
+  }
+
   /**
-   * Calculate the incident light. Will automatically pull from the cache or calculate new values. Cache values are
-   * bilinearly interpolated.
+   * Calculate the incident light. Will automatically pull from the cache or calculate new values.
```

Comment:
Might want to remove `or calculate new values` since the cache is now pregenerated.